src/FakeItEasy/Expressions/ExpressionArgumentConstraintFactory.cs

```diff
@@ -106,5 +116,95 @@ private IArgumentConstraint CreateParamArrayConstraint(NewArrayExpression expres
 
             return new AggregateArgumentConstraint(result);
         }
+
+        private class ArgumentConstraintExpressionVisitor : ExpressionVisitor
+        {
+            private readonly List<Expression> constraintExpressions;

```

Comment:
I just thought it would leave us with slightly simpler code and would be (probably imperceptibly) faster in the (very rare) case where a user makes this error. Your choice.